Warn users about file carving size limits (#47376)
fleetdm/fleet: articles/file-carving.md

# File carving

Fleet supports file carving, which allows you to request files (and sets of files) and their full contents from hosts.

File carving data can be either stored in Fleet's database or to an external S3 bucket. For information on how to configure the latter, consult the [configuration docs](https://fleetdm.com/docs/deploying/configuration#s-3-file-carving-backend).

## Setup

In your agent configuration, add the following [command line flags](https://fleetdm.com/docs/configuration/agent-configuration#options-and-command-line-flags) to enable carving:

```yaml
  disable_carver: false
  carver_disable_function: false
  carver_start_endpoint: /api/v1/osquery/carve/begin
  carver_continue_endpoint: /api/v1/osquery/carve/block
  carver_block_size: 8000000
```

> The configured `carver_block_size` must be less than the value of `max_allowed_packet` in the MySQL connection, allowing for some overhead. The default for [MySQL 8](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) it is 64MB (`67108864`).

Reviewer: ```suggestion
The configured `carver_block_size` must be less than the value of `max_allowed_packet` in the MySQL connection, allowing for some overhead. The default for [MySQL 8](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) is 64MB (`67108864`).
```

For the S3-compatible backend, `carver_block_size` must be set to at least 5MiB (`5242880`) due to the [constraints of S3's multipart
uploads](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html).

Compression of the carve contents can be enabled with the `carver_compression` flag. When used, the carve results will be compressed with [Zstandard](https://facebook.github.io/zstd/) compression.

## Create carves

> File carving can cause significant performance impact if multiple factors are scaled up simultaneously. To avoid overloading your Fleet instance:
> - Target a narrow host set. Avoid running carves against all hosts.
> - Use specific paths. Avoid wildcard paths (e.g. /tmp/* or user home directories) that may match many or large files.

> > - Mind the limits. Individual files must be under 8 GB; each file is transferred in 256 MB blocks (one database row and S3 object per block).

Reviewer 2: @noahtalerman , The warning says "each file is transferred in 256 MB blocks" but 256 MB is the server-side max (maxBlockSize in server/service/carves.go), not the actual transfer size. The recommended carver_block_size in the config example above (see line 16 above) is ~8 MB . Maybe say "256 MB max" or reference the configured block size instead.

> > - Mind the limits. Individual files must be under 8 GB; each file is transferred in 256 MB blocks (one database row and S3 object per block).

Author: ```suggestion
> - Mind the limits. Individual files must be under 8 GB
```

@sharon-fdm thanks! I just removed the section. To your point, we talk about the configured block size above.

# File carving

> Fleet supports file carving which allows you to request files (and sets of files) and their full contents from hosts.

Reviewer 3: ```suggestion
Fleet supports file carving, which allows you to request files (and sets of files) and their full contents from hosts.
```

File carving data can be either stored in Fleet's database or to an external S3 bucket. For information on how to configure the latter, consult the [configuration docs](https://fleetdm.com/docs/deploying/configuration#s-3-file-carving-backend).

## Setup

In your agent configuration, add the following [command line flags](https://fleetdm.com/docs/configuration/agent-configuration#options-and-command-line-flags) to enable carving:

```yaml
  disable_carver: false
  carver_disable_function: false
  carver_start_endpoint: /api/v1/osquery/carve/begin
  carver_continue_endpoint: /api/v1/osquery/carve/block
  carver_block_size: 8000000
```

For the (default) MySQL Backend, the configured `carver_block_size` must be less than the value of

> `max_allowed_packet` in the MySQL connection, allowing for some overhead. The default for [MySQL 8](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) it is 64MB (`67108864`).

Reviewer 3: @noahtalerman, is this supposed to be on a separate line? Looks like a copy/paste bug.

Author: Thanks for catching this!

For the S3-compatible backend, `carver_block_size` must be set to at least 5MiB (`5242880`) due to the

> [constraints of S3's multipart

Reviewer 3: @noahtalerman, same here.

For example, to extract the `/etc/hosts` file on a host with hostname `mac-workstation`:

```sh
fleetctl report --hosts mac-workstation --query 'SELECT * FROM carves WHERE carve = 1 AND path = "/etc/hosts"'
```

Glob syntax is also supported to carve entire directories or more:

```sh
fleetctl report --hosts mac-workstation --query 'SELECT * FROM carves WHERE carve = 1 AND path LIKE "/etc/%%"'
```

## Retrieve carves

List the non-expired (see below) carves with `fleetctl get carves`. Note that carves will not be available through this command until Fleet's agent (fleetd) checks in to the Fleet server with the first of the carve contents. This can take some time from initiation of the carve.

To also retrieve expired carves, use `fleetctl get carves --expired`.

Contents of carves are returned as .tar archives, and compressed if that option is configured.

To download the contents of a carve with ID 3, use

```sh
fleetctl get carve --outfile carve.tar 3
```

It can also be useful to pipe the results directly into the tar command for unarchiving:

```sh
fleetctl get carve --stdout 3 | tar -x
```

## Expiration

> Carve contents remain available for 24 hours after the first data is provided from Fleet's agent (fleetd). After this time, the carve contents are cleaned from the database and the carve is marked as "expired".

Reviewer 3: ```suggestion
Carve contents remain available for 24 hours after the first data is provided from Fleet's agent (fleetd). After this time, the carve contents are cleaned from the database, and the carve is marked as "expired".
```

> The same is not true if S3 is used as the storage backend. In that scenario, it is suggested to set up a [bucket lifecycle configuration](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html) to avoid retaining data in excess. Fleet, in an "eventual consistent" manner (i.e. by periodically performing comparisons), will keep the metadata relative to the files carves in sync with what it is actually available in the bucket.

Reviewer 3: ```suggestion
The same is not true if S3 is used as the storage backend. In that scenario, it is suggested to set up a [bucket lifecycle configuration](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html) to avoid retaining data in excess. Fleet, in an "eventual consistent" manner (i.e., by periodically performing comparisons), will keep the metadata relative to the files carves in sync with what is actually available in the bucket.
```

## Alternative carving backends

#### RustFS

Configure the following:
- `FLEET_S3_ENDPOINT_URL=rustfs_host:port`
- `FLEET_S3_BUCKET=bucket_name`
- `FLEET_S3_SECRET_ACCESS_KEY=your_secret_access_key`
- `FLEET_S3_ACCESS_KEY_ID=access_key_id`
- `FLEET_S3_FORCE_S3_PATH_STYLE=true`
- `FLEET_S3_REGION=localhost` or any non-empty string otherwise Fleet will attempt to derive the region.

> If you're testing file carving locally with the docker-compose environment, the `--dev` flag on Fleet server will automatically point carves to the local RustFS container and write to the `carves-dev` bucket (created automatically) without needing to set additional configuration.

Reviewer 3: ```suggestion
If you're testing file carving locally with the Docker-Compose environment, the `--dev` flag on the Fleet server will automatically point carves to the local RustFS container and write to the `carves-dev` bucket (created automatically) without needing to set additional configuration.
```

Reviewer 3: @noahtalerman, should "docker-compose" be a proper noun?

Author: Just took out mention of Docker entirely.

## Troubleshooting

### Check carve status

You can report on the status of carves through queries to the `carves` table.

You can debug carving problems with:

```sh
fleetctl report --labels 'All Hosts' --query 'SELECT * FROM carves'
```

### Ensure `carver_block_size` is set appropriately

`carver_block_size` is an option that sets the size of each part of a file carve that Fleet's agent (fleetd)

> sends to the Fleet server.

Reviewer 3: @noahtalerman, looks like another copy/paste line break.

When using the MySQL backend (default), this value must be less than the `max_allowed_packet`

> setting in MySQL. If it is too large, MySQL will reject the writes.

Reviewer 3: Same here, and all lines up to 115.

> - Target a narrow host set. Avoid running carves against all hosts.
> - Use specific paths. Avoid wildcard paths (e.g. /tmp/* or user home directories) that may match many or large files.
> - Mind the limits. Individual files must be under 8 GB; each file is transferred in 256 MB blocks (one database row and S3 object per block).
> - Avoid scheduled carves on broad targets. Automations that repeat carves against large host sets compound the load over time.
> The total load scales as the product of: number of hosts × number of paths × number of matching files × average file size. Any one of these can be large in isolation, but all four at once can result in millions of database writes and terabytes of S3 data simultaneously.

File carves are initiated with live reports. Run live report using the `carves` table, providing `carve = 1` along with the desired path(s) as constraints.

For example, to extract the `/etc/hosts` file on a host with hostname `mac-workstation`:

```sh
fleetctl report --hosts mac-workstation --query 'SELECT * FROM carves WHERE carve = 1 AND path = "/etc/hosts"'
```

> The globbing syntax is also supported to carve entire directories or more:

Reviewer: ```suggestion
Glob syntax is also supported to carve entire directories or more:
```

```sh
fleetctl report --hosts mac-workstation --query 'SELECT * FROM carves WHERE carve = 1 AND path LIKE "/etc/%%"'
```

## Retrieve carves

List the non-expired (see below) carves with `fleetctl get carves`. Note that carves will not be available through this command until Fleet's agent (fleetd) checks in to the Fleet server with the first of the carve contents. This can take some time from initiation of the carve.

To also retrieve expired carves, use `fleetctl get carves --expired`.

Contents of carves are returned as .tar archives, and compressed if that option is configured.

To download the contents of a carve with ID 3, use

```sh
fleetctl get carve --outfile carve.tar 3
```

It can also be useful to pipe the results directly into the tar command for unarchiving:

```sh
fleetctl get carve --stdout 3 | tar -x
```

## Expiration

> Carve contents remain available for 24 hours after the first data is provided from the Fleet's agent (fleetd). After this time, the carve contents are cleaned from the database and the carve is marked as "expired".

Reviewer: ```suggestion
Carve contents remain available for 24 hours after the first data is provided from Fleet's agent (fleetd). After this time, the carve contents are cleaned from the database and the carve is marked as "expired".
```

> The same is not true if S3 is used as the storage backend. In that scenario, it is suggested to setup a [bucket lifecycle configuration](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html) to avoid retaining data in excess. Fleet, in an "eventual consistent" manner (i.e. by periodically performing comparisons), will keep the metadata relative to the files carves in sync with what it is actually available in the bucket.

Reviewer: ```suggestion
The same is not true if S3 is used as the storage backend. In that scenario, it is suggested to set up a [bucket lifecycle configuration](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html) to avoid retaining data in excess. Fleet, in an "eventual consistent" manner (i.e. by periodically performing comparisons), will keep the metadata relative to the files carves in sync with what it is actually available in the bucket.
```

## Alternative carving backends

#### RustFS

Configure the following:
- `FLEET_S3_ENDPOINT_URL=rustfs_host:port`
- `FLEET_S3_BUCKET=bucket_name`
- `FLEET_S3_SECRET_ACCESS_KEY=your_secret_access_key`
- `FLEET_S3_ACCESS_KEY_ID=access_key_id`
- `FLEET_S3_FORCE_S3_PATH_STYLE=true`
- `FLEET_S3_REGION=localhost` or any non-empty string otherwise Fleet will attempt to derive the region.

If you're testing file carving locally with the docker-compose environment, the `--dev` flag on Fleet server will automatically point carves to the local RustFS container and write to the `carves-dev` bucket (created automatically) without needing to set additional configuration.

## Troubleshooting

### Check carve status

You can report on the status of carves through queries to the `carves` table.

> The details provided by

Reviewer: ```suggestion
You can debug carving problems with:
```

```sh
fleetctl report --labels 'All Hosts' --query 'SELECT * FROM carves'
```

> can be helpful to debug carving problems.

Reviewer: ```suggestion
```

## Setup

In your agent configuration, add the following [command line flags](https://fleetdm.com/docs/configuration/agent-configuration#options-and-command-line-flags) to enable carving:

```yaml
command_line_flags:
  disable_carver=false
  carver_disable_function=false
  carver_start_endpoint=/api/v1/osquery/carve/begin
  carver_continue_endpoint=/api/v1/osquery/carve/block
  carver_block_size=8000000
```

For the (default) MySQL Backend, the configured `carver_block_size` must be less than the value of
`max_allowed_packet` in the MySQL connection, allowing for some overhead. The default for [MySQL 8](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_max_allowed_packet) it is 64MB.

For the S3-compatible backend, `carver_block_size` must be set to at least 5MiB (`5242880`) due to the
[constraints of S3's multipart
uploads](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html).

Compression of the carve contents can be enabled with the `carver_compression` flag. When used, the carve results will be compressed with [Zstandard](https://facebook.github.io/zstd/) compression.

## Create carves

> 

Author: ```suggestion

> File carving can cause significant performance impact if multiple factors are scaled up simultaneously. To avoid overloading your Fleet instance:
> - Target a narrow host set. Avoid running carves against all hosts.
> - Use specific paths. Avoid wildcard paths (e.g. /tmp/* or user home directories) that may match many or large files.
> - Mind the limits. Individual files must be under 8 GB; each file is transferred in 256 MB blocks (one database row and S3 object per block).
> - Avoid scheduled carves on broad targets. Automations that repeat carves against large host sets compound the load over time.
> The total load scales as the product of: number of hosts × number of paths × number of matching files × average file size. Any one of these can be large in isolation, but all four at once can result in millions of database writes and terabytes of S3 data simultaneously.

```

Author: @rfairburn when you get the chance, can you please help me come up w/ a good warning here?

So `shackleton` and other customers know the best practice max: https://fleetdm.slack.com/archives/C094ZAWGDHC/p1779392158155149

cc @zayhanlon

Reviewer 4: In the case you refer to there were a couple of things that at first glance happened that didn't seem intentional

- _all_ hosts got targeted for the carve instead of a narrow target set
- the paths carved had wildcard paths in places such as /tmp or user homes that can sometime have many large files

The net effect was a load both in bandwidth and db usage that came from an X * Y * Z * A situation where:

X = number of hosts
Y = number of paths requests
Z = number of files in the paths
A = avg file size of any individual file

Any one of these can be bigger in scope usually without there being a huge issue for example:

- you can target a large number of hosts for a single small file
- you can look in several paths if you are trying to find a file/filename that only exists in 1-2 of said paths
- there can be a higher-quantiy group of small files in a single path for a single or handul of hosts
- a single large file (up to 8gb I believe) can easily be qeried

However, when all four of the dials end up being turned up (on accident in this case from the sounds of it), you suddenly end up with the db trying to write millions of new rows of metatada, while fleet is trying to populate 10s of terrabytes of data in the s3 carves bucket, while all hosts are both responding to the distributed query that does the `SELECT` to trigger the carve _and_ push all of this data over the carves endpoints.

Another gotcha that did not occur in this specific case is the carves in question were only a one-off.  There wasn't a scheduled query/automations=on situation, which can also compound.

The only nuance is that a particular carve select with multiple files would get combined into a single tar archive before being transferred, but we would still be subject to the 8GB size, block size of 256mb (row and s3 object per block), and the 4th-power load amplification if all factors get scaled at once.

# File carving

Fleet supports file carving which allows you to request files (and sets of files) and their full contents from hosts.

File carving data can be either stored in Fleet's database or to an external S3 bucket. For information on how to configure the latter, consult the [configuration docs](https://fleetdm.com/docs/deploying/configuration#s-3-file-carving-backend).

## Setup

In your agent configuration, add the following [command line flags](https://fleetdm.com/docs/configuration/agent-configuration#options-and-command-line-flags) to enable carving:

```yaml
command_line_flags:
  disable_carver=false
  carver_disable_function=false
  carver_start_endpoint=/api/v1/osquery/carve/begin

>   carver_continue_endpoint=/api/v1/osquery/carve/block

Reviewer 2: @noahtalerman, Since the format changed from shell flags (`--flag=value`) to YAML, the separator should be `:` not `=`.

```suggestion
  disable_carver: false
  carver_disable_function: false
  carver_start_endpoint: /api/v1/osquery/carve/begin
  carver_continue_endpoint: /api/v1/osquery/carve/block
  carver_block_size: 8000000
```

Author: Thanks!

>   carver_block_size=8000000

Reviewer: @noahtalerman is this intentional or a duplicate?

Author: Whoops! It's not. Fixed that.

>   carver_block_size=8000000

Author: ```suggestion
```

For the S3-compatible backend, `carver_block_size` must be set to at least 5MiB (`5242880`) due to the
[constraints of S3's multipart
uploads](https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html).

Compression of the carve contents can be enabled with the `carver_compression` flag. When used, the carve results will be compressed with [Zstandard](https://facebook.github.io/zstd/) compression.

## Create carves

> File carving can cause significant performance impact if multiple factors are scaled up simultaneously. To avoid overloading your Fleet instance:
> - Target a narrow host set. Avoid running carves against all hosts.
> - Use specific paths. Avoid wildcard paths (e.g. /tmp/* or user home directories) that may match many or large files.
> - Mind the limits. Individual files must be under 8 GB; each file is transferred in 256 MB blocks (one database row and S3 object per block).
> - Avoid scheduled carves on broad targets. Automations that repeat carves against large host sets compound the load over time.

> > The total load scales as the product of: number of hosts × number of paths × number of matching files × average file size. Any one of these can be large in isolation, but all four at once can result in millions of database writes and terabytes of S3 data simultaneously.

Author: @rfairburn what do you think?

Based on your comment here: https://github.com/fleetdm/fleet/pull/47376#discussion_r3392188942
